DBZ-9020 Add documentation for OpenLineage integration (#6462)
debezium/debezium: documentation/modules/ROOT/pages/integrations/openlineage.adoc

* *Data governance and compliance*
* *Impact analysis* when making changes
* *Debugging data quality issues*
* *Understanding data dependencies*

https://openlineage.io/[OpenLineage] is an open standard for data lineage that provides a unified way to collect and track lineage metadata across different data systems. It defines a common model for describing datasets, jobs, and runs, making it easier to build comprehensive lineage graphs across heterogeneous data infrastructures.

For complete information about OpenLineage, visit the https://openlineage.io/[official OpenLineage website] and https://openlineage.io/docs/[documentation].

== How Debezium OpenLineage Integration Works

The Debezium OpenLineage integration maps Debezium components to the OpenLineage data model as follows:

=== OpenLineage Job Mapping

The Debezium connector is mapped to an OpenLineage *Job*, which includes:

* Name will be the connector name
* Namespace will be `topic.prefix` if not specified with `openlineage.integration.job.namespace`
* Debezium version information
* Complete connector configuration

> * Job metadata (description, tags, owners)

Author: > Namespace:: Inherited from `topic.prefix`, if specified, or defaults to `openlineage.integration.job.namespace`.

This is effectively the opposite

2. View the dependency tree for your chosen `openlineage-java` version to identify the exact versions of all transitive dependencies
3. Download all required JAR files and place them in your Debezium connector's classpath

[NOTE]
====
Dependency versions must be compatible with each other. Always refer to the Maven dependency tree of the specific `openlineage-java` version you plan to use to ensure compatibility.
====

== Configuration

=== Basic Configuration

To enable OpenLineage integration, add the following properties to your connector configuration:

[source,properties]
----
# Enable OpenLineage integration
openlineage.integration.enabled=true

# Path to OpenLineage configuration file
openlineage.integration.config.file.path=/path/to/openlineage.yml

# Job metadata (optional but recommended)
openlineage.integration.job.namespace=myNamespace
openlineage.integration.job.description=CDC connector for products database
openlineage.integration.job.tags=env=prod,team=data-engineering
openlineage.integration.job.owners=Alice Smith=maintainer,Bob Johnson=Data Engineer
----

=== OpenLineage Configuration File

> Create an `openlineage.yml` file to configure the OpenLineage client. Here's a basic example:

Author: You need to put in a place where it will be accessible by the Kafka Connect and the put the path in the `openlineage.integration.config.file.path`

key.converter=org.apache.kafka.connect.json.JsonConverter
value.converter=org.apache.kafka.connect.json.JsonConverter
----

== Lineage Events

The integration produces several types of OpenLineage events:

=== Run Events

* *START*: When the connector starts
* *RUNNING*: During normal operation and when processing tables
* *COMPLETE*: When the connector stops gracefully
* *FAIL*: When the connector encounters errors

=== Dataset Information

*Input Datasets* represent source database tables. The namespace follows the https://openlineage.io/docs/spec/naming#dataset-naming[OpenLineage dataset naming specification].

For example, with PostgreSQL:

* Namespace: `postgres://hostname:port`
* Name: `schema.table`
* Schema: Column names and types from the source table

The exact namespace format depends on your database system and follows the OpenLineage specification for dataset naming conventions.

*Output Datasets* represent Kafka topics (when using the OpenLineage transform):

* Namespace: `kafka://bootstrap-server:port`
* Name: `topic-prefix.schema.table`
* Schema: Complete CDC event structure including metadata fields

== Monitoring and Troubleshooting

=== Verifying Integration

> 

Reviewer: ```suggestion
You can perform several tasks to verify that the integration is working as expected.

.Procedure
```

> 1. *Check connector logs* for OpenLineage-related messages
> 2. *Verify events in your OpenLineage backend* (if using HTTP transport)
> 3. *Use console transport* for testing:

Reviewer: ```suggestion
1. Check the connector logs for messages that refer to OpenLineage.
2. Verify events in your OpenLineage backend. 
This applies only if you use HTTP transport.
3. Use console transport for testing, as shown in the following example:
```

+
[source,yaml]
----
transport:
  type: console
----

> === Common Issues

Reviewer: ```suggestion
.Common issues
```

> *Integration not working:*

Reviewer: ```suggestion
Integration not working::
```

> 
> * Verify `openlineage.integration.enabled=true`
> * Check that the configuration file path is correct and accessible
> * Ensure the OpenLineage configuration file is valid YAML
> * Verify all required JAR dependencies are present in the classpath

Reviewer: ```suggestion
* Verify that `openlineage.integration.enabled` is set to `true`.
* Check that the path to the OpenLineage configuration file that is specified in the connector configuration is correct, and that {prodname} can access the target file.
* Ensure that the YAML in the OpenLineage configuration file is valid.
* Verify that all required JAR dependencies are present in the classpath.
```

> *Missing output datasets:*

Reviewer: ```suggestion
Missing output datasets::
```

> * Verify the OpenLineage transform is configured
> * Check that `schema.history.internal.kafka.bootstrap.servers` is set

Reviewer: ```suggestion
* Verify that you configured the connector to use the OpenLineage transformation.
* Check that you set the property `schema.history.internal.kafka.bootstrap.servers` in the connector configuration.
```

> *Connection issues:*
> 
> * Verify OpenLineage server URL and authentication
> * Check network connectivity from Debezium to OpenLineage server

Reviewer: ```suggestion
Connection issues::

* Verify that you specified the correct server URL and authentication information in the OpenLineage client configuration.
* Check the network connectivity between {prodname} and the OpenLineage server.
```

> *Dependency issues:*
> 
> * Ensure all required JAR files are present and compatible versions
> * Check for classpath conflicts with existing dependencies

Reviewer: ```suggestion
Dependency issues::

* Ensure that all required JAR files are present and their versions are compatible versions.
* Check for classpath conflicts with existing dependencies.
```

> === Error Events

Reviewer: ```suggestion
.Error Events
```

> When the connector fails, OpenLineage FAIL events include:

Reviewer: ```suggestion
When the connector fails, check for the following items in OpenLineage FAIL events:
```

// Category: debezium-using
// Type: assembly
// ModuleID: open-lineage-integration
// Title: OpenLineage Integration
[id="open-linegage-integration"]
= OpenLineage Integration

:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highlighter: highlight.js

toc::[]

> {prodname} provides built-in integration with OpenLineage to automatically track data lineage for Change Data Capture (CDC) operations. 

Reviewer 2: @mfvitale Does it make sense to add OLNG support to sink connectors too? If yes then maybe we should think a bit more about docs structure as source and sink connectors will be a bit different. Then it probably makes sense to wite this part as explicitly source connector and have a sink part saying like this is not done yet.

Author: > Does it make sense to add OLNG support to sink connectors too?

I think for `sink` you are referring to JDBC and Mongo right? If yes, then your are right.

Output Datasets::
Represent the Kafka topics that result after you apply the OpenLineage transform.
Mappings are created according to the following principles:

* Each Kafka topic that the connector produces becomes an output dataset.
* The output dataset captures the complete CDC event structure, including metadata fields.
* The name of the dataset is based on the connector's topic prefix configuration.

.Run Events

When you integrate {prodname} with OpenLineage, the connector emits events to report changes of status.
Status information is emitted to the following OpenLineage run events:

START:: Reports connector initialization.
RUNNING:: Emitted periodically during normal streaming operations and during processing individual tables. These periodic events ensure continuous lineage tracking for long-running streaming CDC operations.
COMPLETE:: Reports that the connector shut down gracefully.
FAIL:: Reports that the connector encountered an error.

== Required Dependencies

The OpenLineage integration requires several JAR files that are not bundled with {prodname}. 
You must download the following dependencies and add them to your {prodname} installation:

> You must obtain the following JAR files to use {prodname} with OpenLineage:

Author: Well this was another option that I had in mind but then I proceeded with the easiest at least in this phase. For sure having a dedicated module the I image will be added as scope "provided" and the bundled separately will be the perfect final solution. I created a separate task https://issues.redhat.com/browse/DBZ-9110

Namespace:: Inherited from `openlineage.integration.job.namespace`, if specified; otherwise defaults to `topic.prefix`
Debezium version information::
Complete connector configuration::
Job metadata:: Description, tags, and owners.

.Dataset Mapping for source connectors

The following dataset mappings are possible: 

Input Datasets::
Automatically created for the database tables that Debezium is configured to capture changes from.
The following characteristics affect the mapping: 

* Each monitored table becomes an input dataset.
* Schema information for a table is captured, including column names and types.
* Dataset schemas are updated dynamically after a DDL change.

Output Datasets::
Represent the Kafka topics that result after you apply the OpenLineage transform.
Mappings are created according to the following principles:

* Each Kafka topic that the connector produces becomes an output dataset.
* The output dataset captures the complete CDC event structure, including metadata fields.
* The name of the dataset is based on the connector's topic prefix configuration.

.Dataset Mapping for sink connectors

> The support for sink connectors is not yet available.

Reviewer: ```suggestion
Support for sink connectors is not yet available.
```

.Run Events

When you integrate {prodname} with OpenLineage, the connector emits events to report changes of status.

> Status information is emitted to the following OpenLineage run events:

Reviewer: ```suggestion
The connector emits OpenLineage run events after the following status changes:
```

START:: Reports connector initialization.
RUNNING:: Emitted periodically during normal streaming operations and during processing individual tables. These periodic events ensure continuous lineage tracking for long-running streaming CDC operations.
COMPLETE:: Reports that the connector shut down gracefully.
FAIL:: Reports that the connector encountered an error.

== Required Dependencies

The OpenLineage integration requires several JAR files that are not bundled with {prodname}. 

> You must download the following dependencies and add them to your {prodname} installation:
> 
> 
> You must obtain the following JAR files to use {prodname} with OpenLineage:

Reviewer: ```suggestion
To use {prodname} with OpenLineage, you must download the JAR files for the following dependencies and add them to your {prodname} installation:
```

`jackson-dataformat-yaml-_<version>_.jar`:: YAML parsing support
`jackson-datatype-jsr310-_<version>_.jar`:: Java 8 time API support for Jackson
`micrometer-commons-_<version>_.jar`:: Micrometer metrics commons
`micrometer-core-_<version>_.jar`:: Micrometer metrics core
`snakeyaml-_<version>_.jar`:: YAML parser

== Obtaining Dependencies

The dependencies that are required to support the integration might be available in multiple versions. 
To install the dependencies, determine the version that is required and then download the JAR file.

.Procedure

1. On Maven Central, open the https://mvnrepository.com/artifact/io.openlineage/openlineage-java[`openlineage-java` repository] and identify the version used by {prodname}.
2. View the dependency tree for the `openlineage-java` version and identify the exact versions of all transitive dependencies.
3. Download the required JAR files and place them in the classpath for your {prodname} connector.

[NOTE]
====
Dependency versions must be compatible with each other. 
Always refer to the Maven dependency tree of the specific `openlineage-java` version you plan to use to ensure compatibility.
====

== Configuring the integration

> To  enable the integration, you must configure the {prodname} connector and the OpenLineage client.

Reviewer: ```suggestion
To enable the integration, you must configure the {prodname} connector and the OpenLineage client.
```
